torchrec_dlrm/dlrm_main.py

```diff
@@ -391,98 +351,45 @@ def _train(
-            will be run within an epoch.
-        limit_train_batches (Optional[int]): Number of train batches.
-        limit_val_batches (Optional[int]): Number of validation batches.
-
-
+        lr_scheduler (LRPolicyScheduler): Scheduler to adjust the learning rate.
+        print_lr (bool): Whether to print the learning rate every training step.
+        validation_freq (Optional[int]): The number of training steps between validation runs within an epoch.
+        limit_train_batches (Optional[int]): Limits the training set to the first `limit_train_batches` batches.
+        limit_val_batches (Optional[int]): Limits the validation set to the first `limit_val_batches` batches.
 
     Returns:
         None.
     """
-    train_pipeline._model.train()
-
-    # Because TrainPipelineSparseDist buffer batches internally, we load in
-    # TRAIN_PIPELINE_STAGES - 1 batches from the next_iterator into the buffers so that
-    # when train_val_test switches to the next phase, train_pipeline will start
-    # producing results for the TRAIN_PIPELINE_STAGES - 1 buffered batches (as opposed
-    # to the last TRAIN_PIPELINE_STAGES - 1 batches from iterator).
-    combined_iterator = itertools.chain(
-        iterator
-        if limit_train_batches is None
-        else itertools.islice(iterator, limit_train_batches - (epoch > 0)*(TRAIN_PIPELINE_STAGES - 1)),
-        itertools.islice(next_iterator, TRAIN_PIPELINE_STAGES - 1),
-    )
-    samples_per_trainer_across_epochs = TOTAL_TRAINING_SAMPLES / dist.get_world_size() * epochs
-    samples_per_trainer = TOTAL_TRAINING_SAMPLES / dist.get_world_size()
-    num_batches = samples_per_trainer / batch_size
-    if limit_train_batches is not None:
-        num_batches = min(limit_train_batches, num_batches)
+    pipeline._model.train()
+
+    iterator = itertools.islice(iter(train_dataloader), limit_train_batches)
+    # Two filler batches are appended to the end of the iterator to keep the pipeline active while the
+    # last two remaining batches are still in progress awaiting results.
+    two_filler_batches = itertools.islice(iter(train_dataloader), TRAIN_PIPELINE_STAGES - 1)
+    iterator = itertools.chain(iterator, two_filler_batches)
+
     is_rank_zero = dist.get_rank() == 0
     if is_rank_zero:
-        pbar = tqdm(iter(int, 1), desc=f"Epoch {epoch}", disable=False)
+        pbar = tqdm(iter(int, 1), desc=f"Epoch {epoch}", total=len(train_dataloader), disable=False)
     for it in itertools.count():
         try:
-            if is_rank_zero:
-                if print_lr:
-                    for i, g in enumerate(train_pipeline._optimizer.param_groups):
-                        print(f"lr: {it} {i} {g['lr']:.6f}")
-            train_pipeline.progress(combined_iterator)
+            if validation_freq and it > 0 and it % validation_freq == 0:
+                new_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
```

Comment:
here we will be running full val_pipeline over val_dataloader every validation_freq. Is this intended?

Reply:
is it necessary to create so many pipelines? why not reuse same and just pass in new iters over the dataloaders?

Reply:
It's necessary to create pipelines in this way if we want to keep the code this clean.   Also I'm not seeing impact on performance when comparing the mlperf dcn 1-hot benchmark timings before and after this change.

Reply:
> here we will be running full val_pipeline over val_dataloader every validation_freq. Is this intended?

Yes this is intended.